File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

adjs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

def solution(maps):
    m = len(maps)
    n = len(maps[0])
    visited = [[0 for x in range(n)] for y in range(m)]
    depths = [[0 for x in range(n)] for y in range(m)]
    queue = deque([(0,0)])
    
    while queue:
        current = queue.popleft()
        visited[current[0]][current[1]] = 1
        for x, y in adjs:
            next_x = current[0] + x
            next_y = current[1] + y
            if next_x >= 0 and next_y >= 0 and next_x < m and next_y < n and maps[next_x][next_y] and not visited[next_x][next_y]:
                depths[next_x][next_y] = depths[current[0]][current[1]] + 1
                queue.append((next_x, next_y))
    answer = depths[m-1][n-1]
    if answer == 0:
        return -1
    return answer + 1

def solution(maps):
    m = len(maps)
    n = len(maps[0])
    visited = [[0 for x in range(n)] for y in range(m)]
    depths = [[0 for x in range(n)] for y in range(m)]
    queue = deque([(0,0)])
    
    while queue:
        current = queue.popleft()
        visited[current[0]][current[1]] = 1
        for x, y in adjs:
            next_x = current[0] + x
            next_y = current[1] + y
            if next_x >= 0 and next_y >= 0 and next_x < m and next_y < n and maps[next_x][next_y] and not visited[next_x][next_y]:
                visited[next_x][next_y] = 1
                depths[next_x][next_y] = depths[current[0]][current[1]] + 1
                queue.append((next_x, next_y))
    answer = depths[m-1][n-1]
    if answer == 0:
        return -1
    return answer + 1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[][] D = {{0,1},{1,0},{-1,0},{0,-1}};
    
    public int solution(int[][] maps) {
        int answer = 0;
        int H = maps.length;
        int W = maps[0].length;
        
        Queue<int[]> q = new ArrayDeque<>();
        int[][] visited = new int[H][W];
        
        q.add(new int[] {0, 0});
        visited[0][0] = 1;
        
        while(!q.isEmpty()) {
            int[] current = q.poll();
            int cx = current[0], cy = current[1];
            
            for(int[] d: D) {
                int nx = cx + d[0], ny = cy + d[1];
                if(0 <= nx && nx < H && 0 <= ny && ny < W && visited[nx][ny] == 0 && maps[nx][ny] == 1) {
                    q.add(new int[] {nx, ny});
                    visited[nx][ny] = visited[cx][cy] + 1;
                }
            }
        }
        
        return visited[H-1][W-1] == 0 ? -1 : visited[H-1][W-1];
    }
}